operator/cmd/syncclusterconfig/sync.go

```diff
@@ -234,141 +239,184 @@ type ClusterConfigStatus struct {
+
+		configVersion = int64(result.ConfigVersion)
+		logger.Info("updated cluster configuration", "config_version", result.ConfigVersion, "removed", toRemove)
 	}
 
-	result, err := s.Client.PatchClusterConfig(ctx, upsert, removed)
+	// Compute a hash of configs that might trigger a restart of the cluster.
+	// This incorrectly uses the desired config instead of the one reflected by
+	// the cluster and could technically trigger a restart prior to the config
+	// be set but is historical consistent.
+	// Use of this hash is slated for removal.
+	hashOfConfigsThatNeedRestart, err := hashConfigsThatNeedRestart(desired, schema)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to hash config: %w", err)
 	}
 
-	logger.Info("updated cluster configuration", "config_version", result.ConfigVersion)
 	status, err = s.Client.ClusterConfigStatus(ctx, true)
 	if err != nil {
 		return nil, err
 	}
+
+	needsRestart := false
+	for _, s := range status {
+		configVersion = max(s.ConfigVersion, configVersion)
+		needsRestart = needsRestart && s.Restart
+	}
+
 	return &ClusterConfigStatus{
-		Version:                       int64(result.ConfigVersion),
-		NeedsRestart:                  slices.ContainsFunc(status, func(s rpadmin.ConfigStatus) bool { return s.Restart }),
+		Version:                       configVersion,
+		NeedsRestart:                  needsRestart,
 		PropertiesThatNeedRestartHash: hashOfConfigsThatNeedRestart,
 	}, nil
 }
 
-func (s *Syncer) maybeMergeSuperusers(ctx context.Context, config map[string]any, superusers []string) {
+func (s *Syncer) normalizeConfig(
+	ctx context.Context,
+	schema rpadmin.ConfigSchema,
+	superusers []string,
+	desired map[string]any,
+) map[string]any {
 	logger := log.FromContext(ctx)
+	normalized := maps.Clone(desired)
+
+	if superusers, ok := s.mergeSuperusers(ctx, superusers, desired); ok {
+		// Conversely, if ok is false, we don't need to unset superusers as it
+		// should already be unset.
+		normalized["superusers"] = superusers
+	}
 
-	superusersConfig, ok := config[superusersEntry]
-	if !ok {
-		// we have no superusers configuration, so don't do anything
-		logger.Info("Configuration does not contain a 'superusers' entry. Skipping superusers merge.")
-		return
+	// Redpanda's cluster configs can be aliased. The preferred name is the one
+	// that's listed in the schema (as opposed to being present in an alias
+	// list).
+	for key, schema := range schema {
+		_, hasKey := normalized[key]
+
+		for _, alias := range schema.Aliases {
+			aliased, hasAlias := normalized[alias]
+
+			if !hasKey && hasAlias {
+				logger.Info("renaming aliased config key", "from", alias, "to", key)
+				normalized[key] = aliased
+				delete(normalized, alias)
+			} else if hasKey && hasAlias {
+				logger.Info("dropping aliased key due to non-alias key being provided", "alias", alias, "non-alias", key)
```

Comment:
just for kicks, did you try this directly? from what I saw when this was not alias-aware, I believe that core roughly does this as well?

Reply:
I dropped all the client side code and fully delegated to redpanda. I could very well be overlooking something but it seems largely random which of the keys will be respected. Sadly (or thankfully) there's no property with multiple aliases to really get in the weeds on.

Reply:
@chrisseto just tried to find this in the c++ codebase and looks like it just does the removals second:
https://github.com/redpanda-data/redpanda/blob/a50231e90ffe3a142b79ad8a30442864d5440d9a/src/v/cluster/config_manager.cc#L718

as long as we're normalizing the properties though, we should be ok

Reply:
Just requested changes, b/c it looks like in the latest iteration you dropped the alias merging code in the normalization function. Given that it looks like the issue is just that core does removals after additions, I'm pretty sure we're going to still have an flapping config if we specify an alias and are in declarative mode.